pkg/util/webhooks/tls.go

```diff
@@ -29,3 +69,135 @@ func SetupTLS(caManager ClientCAManager, serverCert tls.Certificate, clientAuth
+				},
+				// Neither the client nor the server should validate anything itself, `VerifyPeerCertificate` is still executed
+				InsecureSkipVerify: true,
+				// XXX: We need to verify the cert ourselves because we don't have DNS or IP on the certs at the moment
+				VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+
+					// impossible with RequireAnyClientCert
+					if len(rawCerts) == 0 {
+						return fmt.Errorf("no client certificate provided.")
+					}
+
+					c, err := x509.ParseCertificate(rawCerts[0])
+					if err != nil {
+						return fmt.Errorf("failed to parse peer certificate: %v", err)
+					}
+					_, err = c.Verify(x509.VerifyOptions{
+						Roots:     certPool,
+						KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+					})
+
+					if c.Subject.CommonName != "kubevirt.io:system:client:virt-handler" {
+						return fmt.Errorf("common name is invalid, expected %s, but got %s", "kubevirt.io:system:client:virt-handler", c.Subject.CommonName)
+					}
+
+					if err != nil {
+						return fmt.Errorf("could not verify peer certificate: %v", err)
+					}
+					return nil
+				},
+				ClientAuth: tls.RequireAndVerifyClientCert,
+			}
+			return config, nil
+		},
+	}
+}
+
+func SetupTLSForVirtHandlerClients(certManager bootstrap.CertificateManager) *tls.Config {
+	return &tls.Config{
+		InsecureSkipVerify: true,
+		GetCertificate: func(info *tls.ClientHelloInfo) (certificate *tls.Certificate, err error) {
+			cert := certManager.Current()
+			if cert == nil {
+				return nil, fmt.Errorf("No server certificate, server is not yet ready to receive traffic")
+			}
+			return cert, nil
+		},
+		GetClientCertificate: func(info *tls.CertificateRequestInfo) (certificate *tls.Certificate, e error) {
+			cert := certManager.Current()
+			if cert == nil {
+				return nil, fmt.Errorf("No client certificate, client is not yet ready to talk to the server")
+			}
+			return cert, nil
+		},
+		GetConfigForClient: func(info *tls.ClientHelloInfo) (config *tls.Config, err error) {
+			certPool := certManager.CABundle()
+			if certPool == nil {
+				return nil, fmt.Errorf("No ca certificate, server is not yet ready to receive traffic")
+			}
+			cert := certManager.Current()
+			if cert == nil {
+				return nil, fmt.Errorf("No server certificate, server is not yet ready to receive traffic")
+			}
+
+			config = &tls.Config{
+				MinVersion: tls.VersionTLS12,
+				ClientCAs:  certPool,
+				GetCertificate: func(info *tls.ClientHelloInfo) (i *tls.Certificate, e error) {
+					return cert, nil
+				},
+				GetClientCertificate: func(info *tls.CertificateRequestInfo) (certificate *tls.Certificate, e error) {
+					return cert, nil
+				},
+				// Neither the client nor the server should validate anything itself, `VerifyPeerCertificate` is still executed
+				InsecureSkipVerify: true,
+				// XXX: We need to verify the cert ourselves because we don't have DNS or IP on the certs at the moment
```

Comment:
I'm not sure this is okay.  If one virt handler is compromised, then wouldn't they essentially all be?

Reply:
Maybe there should be a virt handler CA that is created/managed by the operator and mounted by the virt handlers.  It is rotated frequently by the operator.  When virt handlers start, they create a cert with that CA.  They refresh the cert when the operator refreshes the CA cert.

Reply:
That we have unique certificates per handler is something I see solved in the future with CSRs. For now we would rotate certificates every day. Since the certificates are only valid for one day, the potential vulnerability of all handlers is at least limited to one day.

Reply:
Just realized my suggestion above doesn't help (an attacker could use the ca cert to create a cert for any virt handler).  But I'm not okay with the CSR approach if it means giving one of our components permission to approve CSRs since the CA for CSRs may be the same CA that the kube-apiserver trusts.

This is really a hard problem.  Perhaps this is a decent compromise.  But maybe we should rotate even more frequently.

Reply:
> Just realized my suggestion above doesn't help (an attacker could use the ca cert to create a cert for any virt handler). But I'm not okay with the CSR approach if it means giving one of our components permission to approve CSRs since the CA for CSRs may be the same CA that the kube-apiserver trusts.

It is not possible to properly solve without CSRs, or creating a system which works the same way. The PR does not intend to solve this btw. It just improves the general situation by adding rotations.

Reply:
> Just realized my suggestion above doesn't help (an attacker could use the ca cert to create a cert for any virt handler)

I don't think that ca rotation is necessary in general, since the certificates themselves are short-lived, but maybe I don't get something.

Reply:
>It is not possible to properly solve this without CSRs, or creating a system which works the same way. The PR does not intend to solve this btw. It just improves the general situation by adding rotations.

Yeah, I am in favor of something similar but not CSRs.  And it is something we can solve later but I just want to be sure that this PR is not making things less secure in any way than current.  And I don't think it is because the current virt-handler certs are self signed

Reply:
> I don't think that ca rotation is necessary in general, since the certificates themselves are short-lived, but maybe I don't get something.

Same issue as with server certs, really.  If the CA cert is compromised, and you know about it, maybe you can regenerate all the certs in your system.  But what if you don't?

Reply:
@mhenriks do we still see value at all in the feature, which can rotate certs without ca-renewal (which also regenerates certs  and has no downtime)?

Reply:
Good question, I guess not if all dependencies are updated when the CA cert is rotated.  May be nice to have that flexibility but not sure it has much value.